Float16 optimal formatting

tbkka opened the pull request:
Extend SwiftDtoa to provide optimal formatting for Float16 and use that for `Float16.description` and `Float16.debugDescription`.

Resolves rdar://61414101

tbkka committed:
Float16 optimal formatting

Extend SwiftDtoa to provide optimal formatting for Float16 and use that for `Float16.description` and `Float16.debugDescription`.

Resolves rdar://61414101

tbkka commented:
@swift-ci Please test

tbkka commented:
One interesting nit:  When encoding NaNs, Swift sets the top payload bit if this is a regular "quiet" nan and the next-most-significant payload bit if this is a "signaling" NaN.  For Float, Double, and Float80, SwiftDtoa uses the first of these to distinguish the two cases.  However, that logic doesn't work for Float16.  It appears that when converting Float16 to and from Float, the encoding for NaNs gets altered:  the most significant payload bit is always set, and the next-most-significant is set for signaling NaNs.  I've altered the SwiftDtoa logic accordingly for Float16 only.  I'm not sure whether this encoding change should be considered a bug in our Float16 support or whether SwiftDtoa should be consistently keying on the second-most-significant bit instead of the first.

tbkka commented:
For those interested in arcane details:  Following gdtoa, SwiftDtoa produces the exact form for integral Float16 inputs, instead of the optimal (shortest) form.  For example, the 0x1p13 formats as 8192.0 instead of the optimal 8190.0.  Similarly, the maximum Float16 is 65504.0 (exact) instead of 65500.0 (optimal).  Since these numbers are too small to render in exponential format, it just seems pointless to use the true optimal with trailing zeros when those same digits could be exact.

stephentyrone commented:
Re sNaN: an unfortunate divergence from other types, mostly unavoidable, but not strictly a violation of IEEE 754. You have to *either* print a string that indicates signaling, *or* signal invalid. The conversion routine signals invalid.

tbkka commented:
@swift-ci Please test

LuizZak commented on include/swift/Runtime/SwiftDtoa.h:
```diff
@@ -102,6 +103,14 @@ int swift_decompose_double(double d,
 #endif
 
+#if SWIFT_DTOA_FLOAT16_SUPPORT
+// See swift_decompose_double.  `digits_length` must be at least 9.
```
```diff
- `digits_length` must be at least 9
+ `digits_length` must be at least 5
```

according to the implementation.

tbkka replied:
Updated comment.

stephentyrone commented on stdlib/public/runtime/SwiftDtoa.cpp:
```diff
@@ -624,9 +624,184 @@ int swift_decompose_double(double d,
+    static const int64_t exponentBias = (1 << (exponentBitCount - 1)) - 2; // 14
+
+    // Step 0: Deconstruct the target number
+    // Note: this strongly assumes IEEE 754 binary16 format
```
__fp16 is a clang extension and is always binary16, so this is not an assumption.

tbkka replied:
Clarified the comment.

stephentyrone commented on stdlib/public/runtime/SwiftDtoa.cpp:
```diff
@@ -993,6 +1168,60 @@ static size_t swift_format_constant(char *dest, size_t length, const char *s) {
+            // XXX FIXME: This is different from Float/Double because
+            // Swift's Float16 NaN encoding gets altered before it gets
+            // here, probably because it gets converted to Float and back.
+            const char *signaling = ((raw >> (significandBitCount - 2)) & 1) ? "s" : "";
```
Let's not try to do this. In particular, this will erroneously print some qNaNs as sNaN, which is significantly worse than printing a sNaN as qNaN.

tbkka replied:
I reverted it to use the same detection as Float/Double.  That sacrifices sNaN printing in the short term, but it should DTRT when we get better argument passing for Float16.

stephentyrone reviewed (changes requested):
Let's not try to finesse sNaN, otherwise this looks good.

tbkka committed:
Correct header comment to match implementation

tbkka committed:
Clarify a comment

tbkka committed:
Detect NaN encoding the same way as Float/Double

Also mark the corresponding tests as `expectedFailure` until we can update the
Float16 argument passing convention to not re-encode NaNs.

Explanation: Swift's Float16 support passes Float16s by reencoding them to/from
Float32.  This works well for most purposes but incidentally loses the signaling
marker from any NaN, with a side effect that the print code never actually sees
a true sNaN.  The earlier code here tried to detect sNaN in a different way, but
that approach isn't guaranteed to work so we decided to make this code use the
correct detection logic -- sNaN printing will just be broken until we can get a
better argument passing convention.

stephentyrone commented on test/stdlib/PrintFloat.swift.gyb:
```diff
@@ -623,11 +623,11 @@ PrintTests.test("Printable_Float16") {
+  expectFailure { expectNaN("-snan", -Float16.signalingNaN) }
+  expectFailure { expectNaN("snan(0xff)", Float16(nan: 255, signaling: true)) }
+  expectFailure { expectNaN("-snan(0xff)", -Float16(nan: 255, signaling: true)) }
+  expectFailure { expectNaN("snan(0xff)", Float16(bitPattern: 0x7dff)) }
```
Please add a couple additional test cases to make sure these still print as nan.

tbkka replied:
Done.

tbkka committed:
Add some positive tests for snan printing until the real tests can be enabled

tbkka commented on test/stdlib/PrintFloat.swift.gyb:
```diff
@@ -578,6 +584,145 @@ PrintTests.test("Printable_CDouble") {
+  let lowerBound = -7 // Or -4?  Does ARM support subnormals?
+#else
+  let lowerBound = -7
+#endif
```
@stephentyrone  -- Should we test subnormals here for ARM?

tbkka replied:
Removed the ARM conditional here.  I also noticed that this test was stopping at 10^2 for not apparent reason; I changed it to go up to 10^4.

tbkka commented on test/stdlib/PrintFloat.swift.gyb:
```diff
@@ -578,6 +584,145 @@ PrintTests.test("Printable_CDouble") {
+  expectDescription("3.8e-06", 0x1p-18 as Float16)
+  expectDescription("7.6e-06", 0x1p-17 as Float16)
+  expectDescription("1.526e-05", 0x1p-16  as Float16)
+  expectDescription("3.05e-05", 0x1p-15 as Float16)
```
Similarly, items above here are testing subnormals; are they supported on all our platforms?

stephentyrone replied:
Yes. We only have hardware support for Float16 arithmetic on arm64, and on arm64 we run with subnormals enabled. On all other platforms, we use Float for the arithmetic, and Float16 inputs will never produce a subnormal result in Float.

stephentyrone commented:
I'm satisfied with this now.

tbkka committed:
Remove unnecessary 🤞 ARM checks

tbkka commented:
@swift-ci Please test and merge

tbkka commented:
@swift-ci Please test and merge

tbkka committed:
Comment out these tests that vary across HW platforms

tbkka commented:
@swift-ci Please test and merge

stephentyrone commented on test/stdlib/PrintFloat.swift.gyb:
```diff
@@ -623,12 +623,20 @@ PrintTests.test("Printable_Float16") {
+  expectNaN("snan(0xff)", Float16(bitPattern: 0x7dff))
+  }
+*/
   expectEqual("nan", Float16.signalingNaN.description)
```
You might need to drop these as well. Can we just check that the result contains "nan"?

tbkka replied:
The `description` property always returns the constant string "nan" for any NaN, ignoring payload and flags.  So the `expectEqual` checks here are correct.

Only the `debugDescription` contains any detail about nans.

tbkka commented:
@swift-ci Please test Linux
